### PR TITLE
[OpenACC] Implement `self` clause for compute constructs

### PR DESCRIPTION
**Pull request by erichkeane**
`self` clauses on compute constructs take an optional condition expression. We again limit the implementation to ONLY compute constructs to ensure we get all the rules correct for others.  However, this one will be particularly complicated, as it takes a `var-list` for `update`, so when we get to that construct/clause combination, we need to do that as well.

This patch also furthers uses of the `OpenACCClauses.def` as it became useful while implementing this (as well as some other minor refactors as I went through).

Finally, `self` and `if` clauses have an interaction with each other, if an `if` clause evaluates to `true`, the `self` clause has no effect. While this is intended and can be used 'meaningfully', we are warning on this with a very granular warning, so that this edge case will be noticed by newer users, but can be disabled trivially.

### CODE REVIEW

File: clang/lib/Parse/ParseOpenACC.cpp [hunk 1]
```diff
@@ -835,18 +835,23 @@ Parser::OpenACCClauseParseResult Parser::ParseOpenACCClauseParams(
...
 
-      if (DefKind == OpenACCDefaultClauseKind::Invalid)
+      if (DefKind == OpenACCDefaultClauseKind::Invalid) {
         Diag(DefKindTok, diag::err_acc_invalid_default_clause_kind);
-      else
+        Parens.skipToEnd();
+        return OpenACCCanContinue();
+      } else {
         ParsedClause.setDefaultDetails(DefKind);
+      }
```
Comment (alexey-bataev) [inline clang/lib/Parse/ParseOpenACC.cpp:854]: No need for else here

File: clang/lib/Sema/SemaOpenACC.cpp [hunk 2]
```diff
@@ -160,12 +169,58 @@ SemaOpenACC::ActOnClause(ArrayRef<const OpenACCClause *> ExistingClauses,
...
-    // TODO OpenACC: When we implement 'self', this clauses causes us to
-    // 'ignore' the self clause, so we should implement a warning here.
+    // If the 'if' clause is true, it makes the 'self' clause have no effect,
+    // diagnose that here.
+    // TODO OpenACC: When we add these two to other constructs, we might not
+    // want to warn on this (for example, 'update').
+    const auto *Itr =
+        llvm::find_if(ExistingClauses, [](const OpenACCClause *C) {
+          return C->getClauseKind() == OpenACCClauseKind::Self;
+        });
```
Comment (alexey-bataev) [inline clang/lib/Sema/SemaOpenACC.cpp:179]:
```suggestion
        llvm::find_if(ExistingClauses, IsaPred<OpenACCSelfClause>);
``` 
???

File: clang/lib/Sema/SemaOpenACC.cpp [hunk 3]
```diff
@@ -160,12 +169,58 @@ SemaOpenACC::ActOnClause(ArrayRef<const OpenACCClause *> ExistingClauses,
...
+      return nullptr;
+
+    // If the 'if' clause is true, it makes the 'self' clause have no effect,
+    // diagnose that here.
+    // TODO OpenACC: When we add these two to other constructs, we might not
+    // want to warn on this (for example, 'update').
+    const auto *Itr =
+        llvm::find_if(ExistingClauses, [](const OpenACCClause *C) {
+          return C->getClauseKind() == OpenACCClauseKind::If;
+        });
```
Comment (alexey-bataev) [inline clang/lib/Sema/SemaOpenACC.cpp:214]:
```suggestion
        llvm::find_if(ExistingClauses, IsaPred<OpenACCIfClause>);
```
???

File: clang/lib/Sema/SemaOpenACC.cpp (hunk 2 above)
Comment (erichkeane) [reply clang/lib/Sema/SemaOpenACC.cpp:179]: Cool!  TIL!

File: clang/lib/AST/OpenACCClause.cpp [hunk 4]
```diff
@@ -72,3 +92,9 @@ void OpenACCClausePrinter::VisitDefaultClause(const OpenACCDefaultClause &C) {
 void OpenACCClausePrinter::VisitIfClause(const OpenACCIfClause &C) {
   OS << "if(" << C.getConditionExpr() << ")";
 }
+
+void OpenACCClausePrinter::VisitSelfClause(const OpenACCSelfClause &C) {
+  OS << "self";
+  if (const Expr *CondExpr = C.getConditionExpr())
```
Comment (alexey-bataev) [inline clang/lib/AST/OpenACCClause.cpp:98]: Need to check if Clause.hasConditionExpr() is true?

Comment (erichkeane) [reply clang/lib/AST/OpenACCClause.cpp:98]: No, all that does is check if `CondExpr` is `nullptr`, so this is doing the same thing.  I thought this interface makes sense (that is, 'null' means no condition).

**Review by alexey-bataev — APPROVED** [review]
LG

**Pull request merged**